### ISSUE TITLE
TPM-based encrypted install fails on physical host

### ISSUE THREAD

**Issue by JackMyers001** [siderolabs/talos#7768]
## Bug Report

When installing Talos v1.5.2 on a physical machine with secure boot and TPM-based disk encryption, it may fail, likely due to missing drivers.

### Description

I tried to install Talos on a bare-metal device with secure boot and full disk encryption, following [the latest guide](https://www.talos.dev/v1.5/talos-guides/install/bare-metal-platforms/secureboot/). The only deviation I have made is using version `1.5.2` instead of the guide's `1.5.0` when pulling the containers to build the installation ISO.

After generating a Talos config (including the `tpm-disk-encryption.yaml`) and applying it to the machine running the `metal-amd64-secureboot.iso`, the node will print `[talos] failed to call key handler at slot 0: stat /dev/tpm0: no such file or directory` every few seconds to the console, and part of the installation will fail. Eventually (after 10+ minutes), the node reboots into the new install in an unconfigured state. Trying to apply the config again will result in the same issue, and no service ever comes up.

When I boot the host into an Arch Linux 2023.09.01 installation ISO, I see both a `/dev/tpm0` and a `/dev/tpmrm0` device. When I install Talos *without* TPM disk encryption and run `talosctl -n <host-ip> ls /dev | grep tpm`, there is no output.

I also tried installing Talos on a Proxmox VM (using the same ISO) using the `q35` machine type and the `OVMF` UEFI. I was able to successfully install Talos with full-disk encryption without issue, and the node came fully online very quickly.

I suspect the missing TPM device is due to the Talos kernel is being built without drivers for many physical TPM chips (maybe only virtual?). Looking at an Arch install on another host, there are several `.ko.zst` files listed under `/lib/modules/<kernel-version>/kernel/drivers/char/tpm`, however on Talos the entire `char` folder is missing (kernel version `6.1.51-talos`). *Note: See the end of this issue for the full Arch directory listing.*

The current commit for Talos' kernel config file has the following set (from [line 3127](https://github.com/siderolabs/pkgs/blob/847a9c3bbea9a8c350dc0b2b84d473c2fd23feb2/kernel/build/config-amd64#L3127)):

```
CONFIG_TCG_TPM=y
CONFIG_HW_RANDOM_TPM=y
CONFIG_TCG_TIS_CORE=y
CONFIG_TCG_TIS=y
# CONFIG_TCG_TIS_I2C is not set
# CONFIG_TCG_TIS_I2C_CR50 is not set
# CONFIG_TCG_TIS_I2C_ATMEL is not set
# CONFIG_TCG_TIS_I2C_INFINEON is not set
# CONFIG_TCG_TIS_I2C_NUVOTON is not set
# CONFIG_TCG_NSC is not set
# CONFIG_TCG_ATMEL is not set
# CONFIG_TCG_INFINEON is not set
# CONFIG_TCG_XEN is not set
CONFIG_TCG_CRB=y
# CONFIG_TCG_VTPM_PROXY is not set
# CONFIG_TCG_TIS_ST33ZP24_I2C is not set
# CONFIG_TELCLOCK is not set
# CONFIG_XILLYBUS is not set
# CONFIG_XILLYUSB is not set
```

Both [Arch Linux](https://gitlab.archlinux.org/archlinux/packaging/packages/linux/-/blob/f4548d92121bcef3902ec30b78c818237bc234f2/config#L4626) and Debian 12 have the following:

```
CONFIG_TCG_TPM=y
CONFIG_HW_RANDOM_TPM=y
CONFIG_TCG_TIS_CORE=y
CONFIG_TCG_TIS=y
CONFIG_TCG_TIS_SPI=m
CONFIG_TCG_TIS_SPI_CR50=y
CONFIG_TCG_TIS_I2C=m
CONFIG_TCG_TIS_I2C_CR50=m
CONFIG_TCG_TIS_I2C_ATMEL=m
CONFIG_TCG_TIS_I2C_INFINEON=m
CONFIG_TCG_TIS_I2C_NUVOTON=m
CONFIG_TCG_NSC=m
CONFIG_TCG_ATMEL=m
CONFIG_TCG_INFINEON=m
CONFIG_TCG_XEN=m
CONFIG_TCG_CRB=y
CONFIG_TCG_VTPM_PROXY=m
CONFIG_TCG_TIS_ST33ZP24=m
CONFIG_TCG_TIS_ST33ZP24_I2C=m
CONFIG_TCG_TIS_ST33ZP24_SPI=m
CONFIG_TELCLOCK=m
CONFIG_XILLYBUS_CLASS=m
CONFIG_XILLYBUS=m
CONFIG_XILLYBUS_PCIE=m
CONFIG_XILLYUSB=m
```

The names of these currently-disabled options (e.g. `CONFIG_TCG_TIS_I2C_INFINEON`) line up with the `ko.zst` files located on Arch (e.g. `tpm_infineon.ko.zst`). I suspect that if these were enabled, the Talos kernel would be able to find my host's TPM chip and install without issue.

TL;DR Please enable all TPM drivers in the kernel config.

### Logs

N/A (as AFAIK I can't pull any logs from a broken install which rejects any `talosctl` commands?)

### Environment

- Talos version:

```
Client:
        Tag:         v1.5.2
        SHA:         undefined
        Built:
        Go version:  go1.21.1
        OS/Arch:     linux/amd64
Server:
        NODE:        192.168.179.11
        Tag:         v1.5.2
        SHA:         318c66b9
        Built:
        Go version:  go1.20.8
        OS/Arch:     linux/amd64
        Enabled:     RBAC
```

- Kubernetes version:

```
Client Version: v1.28.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.28.1
```

- Platform:

**Physical host**

```
Erying Polestar i9 ES (yes this is a weird motherboard)
```

**VM (used for testing only)**

```
VM running on Proxmox with q35 machine type and OVMF UEFI
```

### Arch module directory listing

```
$ tree /lib/modules/6.5.3-arch1-1/kernel/drivers/char/tpm
/lib/modules/6.5.3-arch1-1/kernel/drivers/char/tpm
├── st33zp24
│   ├── tpm_st33zp24_i2c.ko.zst
│   ├── tpm_st33zp24.ko.zst
│   └── tpm_st33zp24_spi.ko.zst
├── tpm_atmel.ko.zst
├── tpm_i2c_atmel.ko.zst
├── tpm_i2c_infineon.ko.zst
├── tpm_i2c_nuvoton.ko.zst
├── tpm_infineon.ko.zst
├── tpm_nsc.ko.zst
├── tpm_tis_i2c_cr50.ko.zst
├── tpm_tis_i2c.ko.zst
├── tpm_tis_spi.ko.zst
├── tpm_vtpm_proxy.ko.zst
└── xen-tpmfront.ko.zst

2 directories, 14 files
```

**Comment by frezbo** [comment]
yes, this seems like missing drivers indeed

**Comment by JackMyers001** [comment]
👍 Please let me know if you produce a build of `ghcr.io/siderolabs/imager` with those drivers enabled and I will test on my machine to confirm whether the issue is fixed.

**Comment by frezbo** [comment]
sure, gimme a few hours :sweat_smile:

**Comment by JackMyers001** [comment]
No rush 😄 I'm in UTC+8 so I likely won't be able to test until tomorrow evening local time.

**Comment by frezbo** [comment]
Okay, here you go: `ghcr.io/frezbo/imager:v1.6.0-alpha.0-38-gc5bd0ac5c@sha256:f8b49d280e808ccd7496b9bf274d5cf4f8e7067bcc8529521c97d0a47017b2b5`

**Comment by JackMyers001** [comment]
Unfortunately that hasn't worked. I first tried installing with TPM encryption enabled and got the same spam as before:

```
[talos] failed to call key handler at slot 0: stat /dev/tpm0: no such file or directory
[talos] kubernetes endpoint watch error {"component": "controller-runtime", "controller": "k8s.EndpointController", "error": "failed to list *v1.Endpoints: Get \x5c"https://localhost:6443/api/v1/namespaces/default/endpoints?fieldSelector=metadata.name%3Dkubernetes&limit=500&resourceVersion=0\x5c": dial dcp 192.168.179.10:6443: connect: no route to host"}
```

That Kubernetes line would be repeated (sometimes it would be https://talos.mycooldomain.com instead of https://localhost as I am wanting to also set up VIP). I think that's actually expected during the install though?

I also tried installing without TPM encryption and succeeded like before. The build number in the top left of the tty is `v1.6.0-alpha.0-38.gc5bd0ac5c` so I definitely haven't accidentally flashed the old `v1.5.2` build. Having said that, when booting the installer the build did actually still report as `v1.5.2` but 100% I regenerated the ISO and flashed the new one (an issue I'd had on `v.1.5.2` in regards to getting the Secure Boot keys added to the UEFI has been fixed 😄).

Regardless, on the new install I ran `talosctl -n <my-node-ip> ls /dev` and see no mention of `tpm` or `tpmrm0`. Running `talosctl -n <my-node-ip> ls /sys/class/tpm` also returns no results. Also, I don't see a a `char` folder under ` /lib/modules/6.1.51-talos/` though if I run `talosctl -n <my-node-ip> cat /lib/modules/6.1.51-talos/modules.builtin | grep infineon` I do get the result `kernel/drivers/char/tpm/tpm_infineon.ko`; did you include the drivers directly, rather than as kernel modules? Not that I'd imagine that to make any difference.

Another thing that I noticed (which did also happen on `v1.5.2` and I forgot about) is during early startup (both the installer and once installed), after Linux has fully loaded there's a message that's printed which reads `[talos] TPM device is not available, skipping PCR extension`. Looking at the code, I assume this is also being printed because it can't find the `/dev/tpm0` device...

Just to confirm, this device absolutely does have a TPM 2.0 which appears and works under Arch, and I can do things like `cat /sys/class/tpm/tpm0/tpm_version_major` which returns "TPM 2.0 device" (or similar).

**Comment by JackMyers001** [comment]
Wait I figured it out what the hell

```
$ talosctl -n <my-node-ip> dmesg | grep tpm
192.168.179.11: kern: warning: [2023-09-18T15:22:03.146875176Z]: tpm_crb: probe of MSFT0101:00 failed with error 378
```

Googling that error message reveals several threads from ~mid-August with the exact same error code. Looks like it's a kernel bug that must have been backported to 6.1?

**Comment by frezbo** [comment]
> Googling that error message reveals several threads from ~mid-August with the exact same error code. Looks like it's a kernel bug that must have been backported to 6.1?

that's entirely possible, things takes time :sweat_smile:

**Comment by JackMyers001** [comment]
Agreed 😄. Absolutely no rush but if you are able to build a kernel with the drivers enabled + that patch I linked, I can see if the issue is resolved tomorrow evening/night.

Really appreciate the fast response btw; getting a new build in some 40 minutes is incredible! ❤️

**Comment by frezbo** [comment]
> Agreed 😄. Absolutely no rush but if you are able to build a kernel with the drivers enabled + that patch I linked, I can see if the issue is resolved tomorrow evening/night.
> 
> Really appreciate the fast response btw; getting a new build in some 40 minutes is incredible! ❤️

do note that we don;t carry any kernel patches and will wait for an upstream fix

**Comment by smira** [comment]
Probably this bug: https://github.com/torvalds/linux/commit/8f7f35e5aa6f2182eabcfa3abef4d898a48e9aa8

**Comment by JackMyers001** [comment]
I reinstalled Talos `v.1.5.2` to confirm if I get the same `dmesg` error I was getting in `v1.6.0-alpha.0-38.gc5bd0ac5c`, and sure enough:

```
$ talosctl -n <my-node-ip> dmesg | grep tpm
192.168.179.11: kern: warning: [2023-09-19T13:20:57.23229923Z]: tpm_crb: probe of MSFT0101:00 failed with error 378
```

I then booted Arch to determine which TPM driver my machine uses. Running `readlink -f /sys/class/tpm/tpm/tpm0/device/driver` returned `/sys/bus/acpi/drivers/tpm_crb`, so it turns out that Talos _already had_ the driver for my TPM; it was just bugged 😅.

For this particular system, adding the extra TPM drivers didn't solve anything, however I don't see any reason why the rest of the TPM drivers shouldn't be enabled in the Talos kernel build config?

> do note that we don;t carry any kernel patches and will wait for an upstream fix

Apologies, I could have phrased that better. The patch I linked was merged into the Linux kernel a couple of weeks ago and has been backported to the `6.1` LTS branch since `6.1.53`.

If you could create a build of Talos with kernel `6.1.53` (or `6.1.54`, which was released about three hours ago 😄) I can confirm if Talos installs successfully with TPM encryption on this machine.

**Comment by frezbo** [comment]
> If you could create a build of Talos with kernel 6.1.53 (or 6.1.54, which was released about three hours ago 😄) I can confirm if Talos installs successfully with TPM encryption on this machine.

Talos 1.5.3 would ship with the latest LTS kernel at the time of release

**Comment by frezbo** [comment]
> however I don't see any reason why the rest of the TPM drivers shouldn't be enabled in the Talos kernel build config

We follow a minimal kernel as possible, so unless someone needs it we'll keep it disabled or try to ship as extension.

**Comment by JackMyers001** [comment]
> We follow a minimal kernel as possible, so unless someone needs it we'll keep it disabled or try to ship as extension.

Fair enough. It looks like several of those disabled drivers are for TPM 1.2 hardware anyway, so it likely wouldn't matter if they were enabled.

> Talos 1.5.3 would ship with the latest LTS kernel at the time of release

Looking forward to it! Should I close out this issue now, or should I do it once `v1.5.3` ships and I can confirm everything's working?

**Comment by frezbo** [comment]
keep it open, until 1.5.3

**Comment by JackMyers001** [comment]
Apologies for the slow update. I built an installer for Talos 1.5.3 and got past my previous error, however I received another error: `failed to call key handler at slot 0: failed to parse hash algorithm: unsupported hash algorithm: 18`.

I decided to dig into [the code](https://github.com/siderolabs/talos/blob/0bd1bdd744f68dc42ac64678972fede992a7189e/internal/pkg/secureboot/tpm2/pcr.go#L166C1-L170C4) to see if I could figure out exactly what was going on.

It turns out that "hash algorithm 18" is `TPMAlgSM3256` in the `go-tpm` library. This algorithm is available (though disabled?) in my BIOS, and AFAIK is mostly used in China? I guess that's what you get for buying a weird Chinese motherboard... 🤷‍♂️

The issue stems from the fact that the current implementation will fail if _any_ PCR is unsupported, rather than succeeding if it finds a _valid_ PCR (currently, only SHA256 is supported). I have created and tested [a patch](https://github.com/JackMyers001/talos/commit/4fcb78d2b55cd705e23238ef0d244b8d5822fc23) which fixes this issue, and I was able to successfully install and boot an encrypted install!

I am happy to create a PR if you think this is the correct solution to this issue?

P.S. I am amazed at how easy it was to compile a build of Talos! Your tooling is amazing 😄

**Comment by frezbo** [comment]
> Apologies for the slow update. I built an installer for Talos 1.5.3 and got past my previous error, however I received another error: `failed to call key handler at slot 0: failed to parse hash algorithm: unsupported hash algorithm: 18`.
> 
> I decided to dig into [the code](https://github.com/siderolabs/talos/blob/0bd1bdd744f68dc42ac64678972fede992a7189e/internal/pkg/secureboot/tpm2/pcr.go#L166C1-L170C4) to see if I could figure out exactly what was going on.
> 
> It turns out that "hash algorithm 18" is `TPMAlgSM3256` in the `go-tpm` library. This algorithm is available (though disabled?) in my BIOS, and AFAIK is mostly used in China? I guess that's what you get for buying a weird Chinese motherboard... 🤷‍♂️
> 
> The issue stems from the fact that the current implementation will fail if _any_ PCR is unsupported, rather than succeeding if it finds a _valid_ PCR (currently, only SHA256 is supported). I have created and tested [a patch](https://github.com/JackMyers001/talos/commit/4fcb78d2b55cd705e23238ef0d244b8d5822fc23) which fixes this issue, and I was able to successfully install and boot an encrypted install!
> 
> I am happy to create a PR if you think this is the correct solution to this issue?
> 
> P.S. I am amazed at how easy it was to compile a build of Talos! Your tooling is amazing 😄

there's already a fix for this here: #7811 

will get it merged soon

> P.S. I am amazed at how easy it was to compile a build of Talos! Your tooling is amazing 😄
:heart:

**Comment by JackMyers001** [comment]
> there's already a fix for this here: #7811

Oh that's great! That PR is a better version of what I wrote, so once that's merged in I think this issue should finally be fully resolved. I'll close this out now 😄
